text/0010-trits/0010-trits.md

```diff
@@ -0,0 +1,124 @@
+- sending trits on the network is done by encoding `5` trits to a single byte for throughput efficiency;
+- for UX purposes, we usually encode `3` trits to a single byte because of the nice readable mapping it offers (`9ABCDEFGHIJKLMNOPQRSTUVWXYZ`);
+
+At the time this RFC is written, the following encodings are considered. This list is not exhaustive, some may be dropped, some may be added.
+
+| Name    | Encoding              | Usage                           | RFC |
+| ------- | --------------------- | ------------------------------- | --- |
+| `t1b1`  | 1 trit per byte       | Hashing, tests                  |     |
+| `t3b1`  | 3 trits per byte      | API, UX, tests                  |     |
+| `t27b8` | 27 trits per 8 bytes  | Troika optimization             |     |
+| `t9b2`  | 9 trits per 2 bytes   | Bee internal encoding           |     |
+| `t5b1`  | 5 trits per byte      | Network communication, storage  |     |
+| `ptrit` | Hardware dependant    | PCurl batch hashing             |     |
+
+Because there are many different trits encodings and switching from one to another is often needed, having a trits interface to have a uniform way to interact with trits, whatever the underlying encoding, could be useful.
+
+This interface should offer methods like allocations, conversions and manipulations.
+
+# Detailed design
+
+Any encoding should be convertible to any encoding. If `N` is the number of encodings available, then the number of possible conversions is `N^2` which could grow quickly. While some of these conversions are heavily used (e.g. `t5b1` to/from `t9b2`) and need to be as efficient as possible, some other are rarely or never expected to be used (e.g. `t5b1` to/from `t3b1`).
+
+To allow implementing all conversions without having to focus too much on unnecessary work, we propose the following requirements:
+- a conversion from any encoding to the same encoding is done by copy;
+- a conversion from `t1b1` to any other encoding is efficiently implemented by `From` trait;
+- a conversion from any encoding to `t1b1` is efficiently implemented by `From` trait;
+- a conversion, proven to be heavily used, from any encoding to any other encoding is efficiently implemented by `From` trait with specialization support;
+- any other conversion falls back to a trivial conversion with an intermediary `t1b1` step;
+
+Note that inner structure should not provide any method to public. It should be
+implemented by methods of `Trits<T>` to that specific type. This interface can
+dereference to inner structure and provide more functionality other than general
+methods.
+
+```rust
+pub trait Encoding {
+    fn with_capacity(n: usize) -> Self;
+    fn len(&self) -> usize;
+    ...
+}
+
+pub struct T1B1 {
+    raw: ...,
+}
+
+impl Encoding for T1B1 {
+    ...
+}
+
+pub struct Trits<T: Encoding>{
+    inner: T,
+}
+
+impl<T: Encoding> Trits<T> {
+    pub fn new() -> Self {
+        Self { inner: T::with_capacity(0) }
```

Comment:
I'm going to also answer question above. Trait should have methods that really necessary. If this can complete `new` and `with_capacity` at the same time. Why should we have redundant method in trait?